### PR TITLE
Add a Force flag to Fetch and Clone, mimicking the command-line client

### PR DESCRIPTION
**Pull request by lupine**
See #490

### CODE REVIEW

**Comment by lupine** [comment]
@mcuadros this is currently lacking tests but you get the basic idea. WDYT?

File: remote.go [hunk 1]
```diff
@@ -580,6 +584,18 @@ func (r *Remote) updateLocalReferenceStorage(
 
 			new := plumbing.NewHashReference(spec.Dst(ref.Name()), ref.Hash())
 
+			if ref.IsBranch() && !force {
+				ff, err := isFastForward(r.s, ref.Hash(), ref.Hash())
```
Comment (lupine) [inline remote.go]: Need to find the hash of the local branch

**Comment by mcuadros** [comment]
Looks good!

**Comment by lupine** [comment]
I'm having a little trouble with the tests, actually. Consider:

```go
func (s *RemoteSuite) TestFetchNoForce(c *C) {
       r := newRemote(memory.NewStorage(), &config.RemoteConfig{
               URL: s.GetBasicLocalRepositoryURL(),
       })

       // Can be fast-forwarded
       r.s.SetReference(plumbing.NewReferenceFromStrings(
               "refs/heads/master", "918c48b83bd081e863dbe1b80f8998f058cd8294",
       ))

       s.testFetch(c, r, &FetchOptions{
               RefSpecs: []config.RefSpec{
                       config.RefSpec("+refs/heads/master:refs/remotes/origin/master"),
               },
       }, []*plumbing.Reference{
               plumbing.NewReferenceFromStrings("refs/heads/master", "f7b877701fbf855b44c0a9e86f3fdce2c298b07f"),
               plumbing.NewReferenceFromStrings("refs/remotes/origin/master", "f7b877701fbf855b44c0a9e86f3fdce2c298b07f"),
       })
}
```

I get the error:

```
remote_test.go:193:
    c.Assert(exp.String(), Equals, r.String())
... obtained string = "f7b877701fbf855b44c0a9e86f3fdce2c298b07f refs/heads/master"
... expected string = "918c48b83bd081e863dbe1b80f8998f058cd8294 refs/heads/master"
```

I want to set up the local ref "master" such that the fetch has to update it. I can then set the ref to commits that are (or are not) ancestors and exercise the behaviour that way, but I've not managed to get it working just yet.

I'll look into it again on Thursday, but in the meantime, any tips appreciated!

**Comment by mcuadros** [comment]
You were lucky? Can you elaborate a little more, I don't understand the goal.

**Comment by mcuadros** [comment]
@lupine any update? If not I will try to take a look to it, since we want to release the v4, and this should be included.

**Comment by lupine** [comment]
Sorry @mcuadros, the tests and the safe locking behaviour both prevented me from finishing this, and in the end we went for a "wrap the SSH client" approach over at GitLab.

There's also the question of respecting "+" in refspecs.

If you can finish this off, I'd be super-grateful!

**Comment by mcuadros** [comment]
Over the go-git SSH client?

**Comment by lupine** [comment]
That's right :/ we're doing https://gitlab.com/gitlab-org/gitlab-shell/blob/master/lib/gitlab_projects.rb#L432 instead.

I'd love to get it cleaned up to be more like https://gitlab.com/gitlab-org/gitlab-shell/merge_requests/147 but couldn't quite manage it.

**Comment by mcuadros** [comment]
https://github.com/src-d/go-git/issues/490#issuecomment-327136452

**Pull request closed without merging**